File: scraper.py
from bs4 import BeautifulSoup
import requests
import re
import json
from datetime import datetime
import time

class BettingOddsScraper:
    def __init__(self):
        self.base_url = 'https://www.betexplorer.com/football/'
        self.leagues = {
            'Champions League': 'europe/champions-league/fixtures/',
            'Premier League': 'england/premier-league/fixtures/',
            'La Liga': 'spain/laliga/fixtures/',
            'Ligue 1': 'france/ligue-1/fixtures/',
            'Bundesliga': 'germany/bundesliga/fixtures/',
            'Serie A': 'italy/serie-a/fixtures/',
            'Eredivisie': 'netherlands/eredivisie/fixtures/',
            'Liga Portugal': 'portugal/liga-portugal/fixtures/',
            'Super Lig': 'turkey/super-lig/fixtures/'
        }
        self.headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
    
    def get_page_content(self, league_key):
        try:
            url = self.base_url + self.leagues[league_key]
            response = requests.get(url, headers=self.headers)
            response.raise_for_status()
            return response.text
        except requests.RequestException as e:
            print(f"Error fetching data: {e}")
            return None
    
    def parse_fixtures_from_html(self, html_content):
        soup = BeautifulSoup(html_content, 'html.parser')
        fixtures = []
        
        rows = soup.find_all('tr')
        current_date = None
        
        for row in rows:
            match_link = row.find('a', class_='in-match')
            if match_link:
                fixture_data = self.extract_fixture_data(row, current_date)
                if fixture_data:
                    fixtures.append(fixture_data)
                    if fixture_data.get('datetime'):
                        current_date = fixture_data['datetime']
        
        return fixtures
    
    def extract_fixture_data(self, row, current_date):
        try:
            match_link = row.find('a', class_='in-match')
            if not match_link:
                return None
            
            team_spans = match_link.find_all('span')
            if len(team_spans) < 2:
                return None
            
            home_team = team_spans[0].get_text().strip()
            away_team = team_spans[1].get_text().strip()
            
            datetime_cell = row.find('td', class_='table-main__datetime')
            match_datetime = None
            if datetime_cell and datetime_cell.get_text().strip():
                match_datetime = datetime_cell.get_text().strip()
            elif current_date:
                match_datetime = current_date
            
            odds_cells = row.find_all('td', class_='table-main__odds')
            odds = {}
            
            if len(odds_cells) >= 3:
                home_button = odds_cells[0].find('button')
                if home_button:
                    odds['home_win'] = float(home_button.get('data-odd', 0))
                
                draw_button = odds_cells[1].find('button')
                if draw_button:
                    odds['draw'] = float(draw_button.get('data-odd', 0))
                
                away_button = odds_cells[2].find('button')
                if away_button:
                    odds['away_win'] = float(away_button.get('data-odd', 0))
            
            return {
                'datetime': match_datetime,
                'home_team': home_team,
                'away_team': away_team,
                'odds': odds
            }
        
        except Exception as e:
            print(f"Error extracting fixture data: {e}")
            return None
    
    def scrape_league_fixtures(self, league_key):
        print(f"Scraping {league_key.replace('_', ' ').title()}...")
        
        html_content = self.get_page_content(league_key)
        if not html_content:
            return []
        
        fixtures = self.parse_fixtures_from_html(html_content)
        return fixtures
    
    def scrape_all_leagues(self):
        all_fixtures = {}
        
        for league_key in self.leagues.keys():
            try:
                fixtures = self.scrape_league_fixtures(league_key)
                all_fixtures[league_key] = fixtures
                print(f"Found {len(fixtures)} fixtures for {league_key}")
                #time.sleep(0.1)
                
            except Exception as e:
                print(f"Error scraping {league_key}: {e}")
                all_fixtures[league_key] = []
        
        return all_fixtures
    
    def save_to_json(self, data, filename='fixtures.json'):
        try:
            with open(filename, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
            print(f"Data saved to {filename}")
        except Exception as e:
            print(f"Error saving data: {e}")
    
    def print_fixtures(self, fixtures, league_name=""):
        if league_name:
            print(f"\n=== {league_name.upper()} ===")
        
        for fixture in fixtures:
            print(f"\n🕐 {fixture.get('datetime', 'TBD')}")
            print(f"⚽ {fixture['home_team']} vs {fixture['away_team']}")
            
            odds = fixture.get('odds', {})
            if odds:
                print(f"📊 Odds - Home: {odds.get('home_win', 'N/A')}, "
                      f"Draw: {odds.get('draw', 'N/A')}, "
                      f"Away: {odds.get('away_win', 'N/A')}")


def parse_existing_html_data(html_string):
    html_cleaned = html_string.strip('[]')
    html_parts = html_cleaned.split('>, <')
    
    fixtures = []
    
    for i, part in enumerate(html_parts):
        if 'in-match' in part and 'href=' in part:
            try:
                if not part.startswith('<'):
                    part = '<' + part
                if not part.endswith('>'):
                    part = part + '>'
                
                soup = BeautifulSoup(part, 'html.parser')
                
                spans = soup.find_all('span')
                if len(spans) >= 2:
                    home_team = spans[0].get_text()
                    away_team = spans[1].get_text()
                    
                    odds_data = {}
                    if i + 1 < len(html_parts):
                        next_parts = html_parts[i:i+4]
                        for odds_part in next_parts:
                            if 'data-odd=' in odds_part:
                                match = re.search(r'data-odd="([\d.]+)"', odds_part)
                                if match:
                                    odds_value = float(match.group(1))
                                    if 'home_win' not in odds_data:
                                        odds_data['home_win'] = odds_value
                                    elif 'draw' not in odds_data:
                                        odds_data['draw'] = odds_value
                                    elif 'away_win' not in odds_data:
                                        odds_data['away_win'] = odds_value
                    
                    fixture = {
                        'home_team': home_team,
                        'away_team': away_team,
                        'odds': odds_data
                    }
                    fixtures.append(fixture)
                    
            except Exception as e:
                continue
    
    return fixtures


if __name__ == "__main__":
    scraper = BettingOddsScraper()
    
    all_data = scraper.scrape_all_leagues()
    scraper.save_to_json(all_data, 'fixtures.json')
    
    for league, fixtures in all_data.items():
        scraper.print_fixtures(fixtures, league)

File: javascript/scraper.js
const puppeteer = require('puppeteer');
const fs = require('fs').promises;

class BettingOddsScraper {
    constructor() {
        this.baseUrl = 'https://www.betexplorer.com/football/';
        this.leagues = {
            'Champions League': 'europe/champions-league/fixtures/',
            'Premier League': 'england/premier-league/fixtures/',
            'La Liga': 'spain/laliga/fixtures/',
            'Ligue 1': 'france/ligue-1/fixtures/',
            'Bundesliga': 'germany/bundesliga/fixtures/',
            'Serie A': 'italy/serie-a/fixtures/',
            'Eredivisie': 'netherlands/eredivisie/fixtures/',
            'Liga Portugal': 'portugal/liga-portugal/fixtures/',
            'Super Lig': 'turkey/super-lig/fixtures/'
        };
        this.browser = null;
        this.page = null;
    }

    async initBrowser() {
        try {
            this.browser = await puppeteer.launch({
                headless: true,
                args: [
                    '--no-sandbox',
                    '--disable-setuid-sandbox',
                    '--disable-blink-features=AutomationControlled'
                ]
            });
            this.page = await this.browser.newPage();
            
            await this.page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36');
            await this.page.setViewport({ width: 1920, height: 1080 });
            
        } catch (error) {
            console.error(`Error initializing browser: ${error}`);
            throw error;
        }
    }

    async closeBrowser() {
        if (this.browser) {
            await this.browser.close();
        }
    }

    async getPageContent(leagueKey) {
        try {
            const url = this.baseUrl + this.leagues[leagueKey];
            console.log(`Navigating to: ${url}`);
            
            await this.page.goto(url, { 
                waitUntil: 'networkidle2',
                timeout: 30000 
            });
            
            await this.page.waitForSelector('tr', { timeout: 10000 });
            
            return await this.page.content();
            
        } catch (error) {
            console.error(`Error fetching data for ${leagueKey}: ${error}`);
            return null;
        }
    }

    async parseFixturesFromPage(leagueKey) {
        try {
            const url = this.baseUrl + this.leagues[leagueKey];
            await this.page.goto(url, { 
                waitUntil: 'networkidle2',
                timeout: 30000 
            });

            try {
                await this.page.waitForSelector('tr', { timeout: 5000 });
            } catch (timeoutError) {
                console.log(`No fixtures found for ${leagueKey} (no table rows present)`);
                return [];
            }

            const fixtures = await this.page.evaluate(() => {
                const rows = document.querySelectorAll('tr');
                const fixturesData = [];
                let currentDate = null;

                rows.forEach(row => {
                    const matchLink = row.querySelector('a.in-match');
                    if (matchLink) {
                        const fixtureData = this.extractFixtureDataFromRow(row, currentDate);
                        if (fixtureData) {
                            fixturesData.push(fixtureData);
                            if (fixtureData.datetime) {
                                currentDate = fixtureData.datetime;
                            }
                        }
                    }
                });

                return fixturesData;
            });

            return fixtures;

        } catch (error) {
            console.error(`Error parsing fixtures for ${leagueKey}: ${error}`);
            return [];
        }
    }

    static extractFixtureDataFromRow(row, currentDate) {
        try {
            const matchLink = row.querySelector('a.in-match');
            if (!matchLink) return null;

            const teamSpans = matchLink.querySelectorAll('span');
            if (teamSpans.length < 2) return null;

            const homeTeam = teamSpans[0].textContent.trim();
            const awayTeam = teamSpans[1].textContent.trim();

            const datetimeCell = row.querySelector('td.table-main__datetime');
            let matchDatetime = null;
            if (datetimeCell && datetimeCell.textContent.trim()) {
                matchDatetime = datetimeCell.textContent.trim();
            } else if (currentDate) {
                matchDatetime = currentDate;
            }

            const oddsCells = row.querySelectorAll('td.table-main__odds');
            const odds = {};

            if (oddsCells.length >= 3) {
                const homeButton = oddsCells[0].querySelector('button');
                if (homeButton) {
                    const homeOdd = homeButton.getAttribute('data-odd');
                    if (homeOdd) odds.home_win = parseFloat(homeOdd);
                }

                const drawButton = oddsCells[1].querySelector('button');
                if (drawButton) {
                    const drawOdd = drawButton.getAttribute('data-odd');
                    if (drawOdd) odds.draw = parseFloat(drawOdd);
                }

                const awayButton = oddsCells[2].querySelector('button');
                if (awayButton) {
                    const awayOdd = awayButton.getAttribute('data-odd');
                    if (awayOdd) odds.away_win = parseFloat(awayOdd);
                }
            }

            return {
                datetime: matchDatetime,
                home_team: homeTeam,
                away_team: awayTeam,
                odds: odds
            };

        } catch (error) {
            console.error(`Error extracting fixture data: ${error}`);
            return null;
        }
    }

    async scrapeLeagueFixtures(leagueKey) {
        console.log(`Scraping ${leagueKey.replace('_', ' ')}...`);
        
        try {
            const url = this.baseUrl + this.leagues[leagueKey];
            await this.page.goto(url, { 
                waitUntil: 'networkidle2',
                timeout: 30000 
            });

            try {
                await this.page.waitForSelector('tr', { timeout: 5000 });
            } catch (timeoutError) {
                console.log(`No fixtures found for ${leagueKey} (no table rows present)`);
                return [];
            }

            const fixtures = await this.page.evaluate(() => {
                const rows = document.querySelectorAll('tr');
                const fixturesData = [];
                let currentDate = null;

                const extractFixtureDataFromRow = (row, currentDate) => {
                    try {
                        const matchLink = row.querySelector('a.in-match');
                        if (!matchLink) return null;

                        const teamSpans = matchLink.querySelectorAll('span');
                        if (teamSpans.length < 2) return null;

                        const homeTeam = teamSpans[0].textContent.trim();
                        const awayTeam = teamSpans[1].textContent.trim();

                        const datetimeCell = row.querySelector('td.table-main__datetime');
                        let matchDatetime = null;
                        if (datetimeCell && datetimeCell.textContent.trim()) {
                            matchDatetime = datetimeCell.textContent.trim();
                        } else if (currentDate) {
                            matchDatetime = currentDate;
                        }

                        const oddsCells = row.querySelectorAll('td.table-main__odds');
                        const odds = {};

                        if (oddsCells.length >= 3) {
                            const homeButton = oddsCells[0].querySelector('button');
                            if (homeButton) {
                                const homeOdd = homeButton.getAttribute('data-odd');
                                if (homeOdd) odds.home_win = parseFloat(homeOdd);
                            }

                            const drawButton = oddsCells[1].querySelector('button');
                            if (drawButton) {
                                const drawOdd = drawButton.getAttribute('data-odd');
                                if (drawOdd) odds.draw = parseFloat(drawOdd);
                            }

                            const awayButton = oddsCells[2].querySelector('button');
                            if (awayButton) {
                                const awayOdd = awayButton.getAttribute('data-odd');
                                if (awayOdd) odds.away_win = parseFloat(awayOdd);
                            }
                        }

                        return {
                            datetime: matchDatetime,
                            home_team: homeTeam,
                            away_team: awayTeam,
                            odds: odds
                        };

                    } catch (error) {
                        console.error(`Error extracting fixture data: ${error}`);
                        return null;
                    }
                };

                rows.forEach(row => {
                    const matchLink = row.querySelector('a.in-match');
                    if (matchLink) {
                        const fixtureData = extractFixtureDataFromRow(row, currentDate);
                        if (fixtureData) {
                            fixturesData.push(fixtureData);
                            if (fixtureData.datetime) {
                                currentDate = fixtureData.datetime;
                            }
                        }
                    }
                });

                return fixturesData;
            });

            return fixtures;

        } catch (error) {
            console.error(`Error scraping ${leagueKey}: ${error}`);
            return [];
        }
    }

    async scrapeAllLeagues() {
        const allFixtures = {};
        
        for (const leagueKey of Object.keys(this.leagues)) {
            try {
                const fixtures = await this.scrapeLeagueFixtures(leagueKey);
                allFixtures[leagueKey] = fixtures;
                console.log(`Found ${fixtures.length} fixtures for ${leagueKey}`);
                
            } catch (error) {
                console.error(`Error scraping ${leagueKey}: ${error}`);
                allFixtures[leagueKey] = [];
            }
        }
        
        return allFixtures;
    }

    async saveToJson(data, filename = 'fixtures.json') {
        try {
            await fs.writeFile(filename, JSON.stringify(data, null, 2), 'utf8');
            console.log(`Data saved to ${filename}`);
        } catch (error) {
            console.error(`Error saving data: ${error}`);
        }
    }

    printFixtures(fixtures, leagueName = "") {
        if (leagueName) {
            console.log(`\n=== ${leagueName.toUpperCase()} ===`);
        }
        
        fixtures.forEach(fixture => {
            console.log(`\n🕐 ${fixture.datetime || 'TBD'}`);
            console.log(`⚽ ${fixture.home_team} vs ${fixture.away_team}`);
            
            const odds = fixture.odds || {};
            if (Object.keys(odds).length > 0) {
                console.log(`📊 Odds - Home: ${odds.home_win || 'N/A'}, ` +
                          `Draw: ${odds.draw || 'N/A'}, ` +
                          `Away: ${odds.away_win || 'N/A'}`);
            }
        });
    }
}

function parseExistingHtmlData(htmlString) {
    const cheerio = require('cheerio');
    const fixtures = [];
    
    try {
        const $ = cheerio.load(htmlString);
        
        $('a.in-match').each((index, element) => {
            try {
                const spans = $(element).find('span');
                if (spans.length >= 2) {
                    const homeTeam = $(spans[0]).text().trim();
                    const awayTeam = $(spans[1]).text().trim();
                    
                    const oddsData = {};
                    const row = $(element).closest('tr');
                    const oddsButtons = row.find('button[data-odd]');
                    
                    if (oddsButtons.length >= 3) {
                        oddsData.home_win = parseFloat($(oddsButtons[0]).attr('data-odd')) || null;
                        oddsData.draw = parseFloat($(oddsButtons[1]).attr('data-odd')) || null;
                        oddsData.away_win = parseFloat($(oddsButtons[2]).attr('data-odd')) || null;
                    }
                    
                    const fixture = {
                        home_team: homeTeam,
                        away_team: awayTeam,
                        odds: oddsData
                    };
                    
                    fixtures.push(fixture);
                }
            } catch (error) {
            }
        });
        
    } catch (error) {
        console.error(`Error parsing HTML data: ${error}`);
    }
    
    return fixtures;
}

async function main() {
    const scraper = new BettingOddsScraper();
    
    try {
        await scraper.initBrowser();
        
        const allData = await scraper.scrapeAllLeagues();
        await scraper.saveToJson(allData, 'fixtures.json');
        
        for (const [league, fixtures] of Object.entries(allData)) {
            scraper.printFixtures(fixtures, league);
        }
        
    } catch (error) {
        console.error(`Main execution error: ${error}`);
    } finally {
        await scraper.closeBrowser();
    }
}

module.exports = {
    BettingOddsScraper,
    parseExistingHtmlData,
    main
};

if (require.main === module) {
    main().catch(console.error);
}